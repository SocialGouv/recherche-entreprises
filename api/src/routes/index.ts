import Router from "koa-router";
import yamljs from "yamljs";

import pkg from "../../package.json";
import { search, searchEntreprise, searchEtablissement } from "../lib";

export const router = new Router();

export const API_PREFIX = "/api/v1";

const parseBoolean = (param: string, defaultz = true) =>
  param === undefined ? defaultz : param.toLowerCase() !== "false";

router.get(`${API_PREFIX}/search`, async (ctx) => {
  const { query, address, limit, onlyWithConvention, open, employer } =
    ctx.query;

  if (!query) {
    ctx.throw(400, `query parameter query is required`);
  }

  try {
    const entreprises = await search({
      addAllConventions: true,
      address: address as string,
      employer: parseBoolean(employer as string, false),
      limit: parseInt(limit as string),
      onlyWithConvention: parseBoolean(onlyWithConvention as string, false),
      open: parseBoolean(open as string, true),
      query: query as string,
    });
    ctx.body = { entreprises };
  } catch (err) {
    console.log(JSON.stringify(err));
    ctx.throw(500);
  }
});

const siretRE = new RegExp(/^\d{14}$/);
const validateSiretFormat = (siret: string) => siret.match(siretRE);

router.get(`${API_PREFIX}/etablissement/:siret`, async (ctx) => {
  const { siret } = ctx.params;

  if (!siret) {
    ctx.throw(400, "siret required");
  } else if (!validateSiretFormat(siret)) {
    ctx.throw(400, "invalid siret");
  }

  try {
    const etablissement = await searchEtablissement(siret as string);
    if (etablissement) {
      ctx.body = etablissement;
    } else {
      ctx.throw(404);
    }
  } catch (err) {
    if ((err as any).status) {
      throw err;
    } else {
      console.log(JSON.stringify(err));
      ctx.throw(500);
    }
  }
});

const sirenRE = new RegExp(/^\d{9}$/);
const validateSirenFormat = (siren: string) => siren.match(sirenRE);

router.get(`${API_PREFIX}/entreprise/:siren`, async (ctx) => {
  const { siren } = ctx.params;

  if (!siren) {
    ctx.throw(400, "siren required");
  } else if (!validateSirenFormat(siren)) {
    ctx.throw(400, "invalid siren");
  }

  try {
    const entreprise = await searchEntreprise(siren as string);
    if (entreprise) {
      ctx.body = entreprise;
    } else {
      ctx.throw(404);
    }
  } catch (err) {
    if ((err as any).status) {
      throw err;
    } else {
      console.log(JSON.stringify(err));
      ctx.throw(500);
    }
  }
});

// kubernetes probe
router.get(`/healthz`, (ctx) => {
  ctx.body = { hello: "world" };
});

router.get(`/version`, (ctx) => {
  ctx.body = {
    about: "https://github.com/SocialGouv/recherche-entreprises",
    success: true,
    version: pkg.version,
  };
});

const spec = yamljs.load("./openapi.yaml");

router.get("/swagger.json", (ctx) => {
  ctx.body = spec;
});
