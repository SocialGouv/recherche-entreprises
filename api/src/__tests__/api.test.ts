import * as http from "http";
import Koa from "koa";
import supertest from "supertest";

import { API_PREFIX, router } from "../routes";

const app = new Koa();
app.use(router.routes());
// .on("error", (err, ctx) => {
// console.error("server error", err, ctx);
// });

const apptest = supertest(http.createServer(app.callback()));

const searchCall = (
  query: string,
  address: string | undefined,
  limit: number | undefined
) => {
  const addressQP = address ? `&a=${address}` : "";
  const limitQP = limit ? `&l=${limit}` : "";

  return apptest.get(
    `${API_PREFIX}/search?onlyWithConvention=true&q=${query}${addressQP}${limitQP}`
  );
};

const michelinSiren = "855200507";
const michelinSiret = `${michelinSiren}03169`;

describe("Test search", () => {
  test("generic search", async () => {
    const { status, body } = await searchCall("michelin", undefined, undefined);
    expect(status).toBe(200);
    expect(body.entreprises).toBeDefined();
    expect(body.entreprises.length).toEqual(20);
    expect(body.entreprises[0].siren).toEqual(michelinSiren);
    // eslint-disable-next-line no-unused-vars
    const { matchingEtablissement, ...partialBody } = body.entreprises[0];
    expect(partialBody).toMatchSnapshot();
  });

  test("with limit", async () => {
    const limit = 50;
    const { body } = await searchCall("michelin", undefined, limit);
    expect(body.entreprises.length).toEqual(limit);
  });

  test("search with postal code and city", async () => {
    const { body: b1 } = await searchCall("michelin", undefined, undefined);
    expect(b1.entreprises[0].matchingEtablissement.address).not.toBe(
      `"23 Place des Carmes Dechaux 63000 Clermont-Ferrand"`
    );

    const { body: b2 } = await searchCall("michelin", "63 000", undefined);
    expect(
      b2.entreprises[0].matchingEtablissement.address
    ).toMatchInlineSnapshot(
      `"23 Place des Carmes Dechaux 63000 Clermont-Ferrand"`
    );

    const { body: b3 } = await searchCall("michelin", "clermont", undefined);
    expect(
      b3.entreprises[0].matchingEtablissement.address
    ).toMatchInlineSnapshot(
      `"23 Place des Carmes Dechaux 63000 Clermont-Ferrand"`
    );

    const { body: b4 } = await searchCall("michelin", "63", undefined);
    expect(
      b4.entreprises[0].matchingEtablissement.address
    ).toMatchInlineSnapshot(
      `"23 Place des Carmes Dechaux 63000 Clermont-Ferrand"`
    );
  });

  test("search with daitrics", async () => {
    const { body: b1 } = await searchCall("michelin", undefined, undefined);
    const { body: b2 } = await searchCall("michélin", undefined, undefined);
    const { body: b3 } = await searchCall("Mîchèlin", undefined, undefined);

    expect(b1).toStrictEqual(b2);
    expect(b1).toStrictEqual(b3);
  });

  test("search with siret", async () => {
    const { body } = await searchCall(michelinSiret, undefined, undefined);
    expect(body.entreprises[0].matchingEtablissement.siret).toEqual(
      michelinSiret
    );
  });
});

describe("Test etablissement search", () => {
  test("correct siret", async () => {
    const { body, status } = await apptest.get(
      `${API_PREFIX}/etablissement/${michelinSiret}`
    );
    expect(status).toEqual(200);
    expect(body.siret).toEqual(michelinSiret);
    expect(body).toMatchSnapshot();
  });

  test("unexisting siret", async () => {
    const { status } = await apptest.get(
      `${API_PREFIX}/etablissement/11111111111111`
    );
    expect(status).toEqual(404);
  });

  test("incorrect siret", async () => {
    const { status } = await apptest.get(
      `${API_PREFIX}/etablissement/${michelinSiret + "123"}`
    );
    expect(status).toEqual(400);
  });
});

describe("Test entreprise search", () => {
  test("correct siren", async () => {
    const { body, status } = await apptest.get(
      `${API_PREFIX}/entreprise/${michelinSiren}`
    );
    // We delete matchingEtablissement since it comes from collapse wich is non deterministic
    delete body.matchingEtablissement;
    expect(status).toEqual(200);
    expect(body.siren).toEqual(michelinSiren);
    expect(body).toMatchSnapshot();
  });

  test("unexisting siret", async () => {
    const { status } = await apptest.get(`${API_PREFIX}/entreprise/111111111`);
    expect(status).toEqual(404);
  });

  test("incorrect siret", async () => {
    const { status } = await apptest.get(
      `${API_PREFIX}/entreprise/${michelinSiren + "123"}`
    );
    expect(status).toEqual(400);
  });
});
