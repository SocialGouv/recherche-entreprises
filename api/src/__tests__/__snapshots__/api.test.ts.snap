// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test entreprise search correct siren 1`] = `
Object {
  "activitePrincipale": "Fabrication et rechapage de pneumatiques",
  "conventions": Array [
    Object {
      "etat": "VIGUEUR_ETEN",
      "id": "KALICONT000005635597",
      "idcc": 45,
      "mtime": 1562873918,
      "shortTitle": "Caoutchouc",
      "texte_de_base": "KALITEXT000005673838",
      "title": "Convention collective nationale du caoutchouc du 6 mars 1953. Étendue par arrêté du 29 mai 1969 JORF 18 juin 1969 ",
      "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635597",
    },
    Object {
      "idcc": 9999,
      "shortTitle": "",
    },
    Object {
      "etat": "VIGUEUR_ETEN",
      "id": "KALICONT000005635173",
      "idcc": 1486,
      "mtime": 1556652289,
      "shortTitle": "Bureaux d'études techniques, cabinets d'ingénieurs-conseils et sociétés de conseils",
      "texte_de_base": "KALITEXT000005679895",
      "title": "Convention collective nationale des bureaux d'études techniques, des cabinets d'ingénieurs-conseils et des sociétés de conseils du 15 décembre 1987. ",
      "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635173",
    },
  ],
  "etablissements": 35,
  "highlightLabel": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "label": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "matching": 35,
  "simpleLabel": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "siren": "855200507",
}
`;

exports[`Test etablissement search correct siret 1`] = `
Object {
  "activitePrincipale": "Traitement de données, hébergement et activités connexes",
  "address": "107 Rue Servient 69003 Lyon",
  "convention": Object {
    "etat": "VIGUEUR_ETEN",
    "id": "KALICONT000005635597",
    "idcc": 45,
    "mtime": 1562873918,
    "shortTitle": "Caoutchouc",
    "texte_de_base": "KALITEXT000005673838",
    "title": "Convention collective nationale du caoutchouc du 6 mars 1953. Étendue par arrêté du 29 mai 1969 JORF 18 juin 1969 ",
    "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635597",
  },
  "etablissements": 35,
  "highlightLabel": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "label": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "matching": 1,
  "simpleLabel": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "siren": "855200507",
  "siret": "85520050703169",
}
`;

exports[`Test search generic search 1`] = `
Object {
  "activitePrincipale": "Fabrication et rechapage de pneumatiques",
  "conventions": Array [
    Object {
      "etat": "VIGUEUR_ETEN",
      "id": "KALICONT000005635597",
      "idcc": 45,
      "mtime": 1562873918,
      "shortTitle": "Caoutchouc",
      "texte_de_base": "KALITEXT000005673838",
      "title": "Convention collective nationale du caoutchouc du 6 mars 1953. Étendue par arrêté du 29 mai 1969 JORF 18 juin 1969 ",
      "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635597",
    },
    Object {
      "etat": "VIGUEUR_ETEN",
      "id": "KALICONT000005635173",
      "idcc": 1486,
      "mtime": 1556652289,
      "shortTitle": "Bureaux d'études techniques, cabinets d'ingénieurs-conseils et sociétés de conseils",
      "texte_de_base": "KALITEXT000005679895",
      "title": "Convention collective nationale des bureaux d'études techniques, des cabinets d'ingénieurs-conseils et des sociétés de conseils du 15 décembre 1987. ",
      "url": "https://www.legifrance.gouv.fr/affichIDCC.do?idConvention=KALICONT000005635173",
    },
  ],
  "etablissements": 35,
  "highlightLabel": "MANUFACTURE FRANCAISE DES PNEUMATIQUES <b><u>MICHELIN</b></u>",
  "label": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "matching": 20,
  "simpleLabel": "MANUFACTURE FRANCAISE DES PNEUMATIQUES MICHELIN",
  "siren": "855200507",
}
`;
